### PR TITLE
[Merged by Bors] - feat: use expected type in set with

### PR DESCRIPTION
**Pull request by alexjbest**
Previously the using `set ... with` line could fail when the given term was ambiguous without the expected type, this was noticed and worked around in #4912.

Zulip discussion: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Timeout.20in.20.60set.20.2E.2E.20with.60/near/364922941

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alexjbest** [comment]
!bench

File: Mathlib/Tactic/Set.lean [hunk 1]
```diff
@@ -39,25 +39,24 @@ h2 : x = y
...
-    evalTactic (← `(tactic| have $h : $val = $a := rfl))
-  | _, _ => pure ()
+    Term.addTermInfo' (isBinder := true) a (mkFVar fvar)
+    if rw.isNone then
+      evalTactic (← `(tactic| try rewrite [(id rfl : $val = $a)] at *))
+    match h, rev with
+    | some h, some none =>
+      evalTactic (← `(tactic| have%$tk $h : $a = ($val : $(← delab ty)) := rfl))
+    | some h, some (some _) =>
+      evalTactic (← `(tactic| have%$tk $h : ($val : $(← delab ty)) = $a := rfl))
```
Comment (alexjbest) [inline Mathlib/Tactic/Set.lean:61]:
```suggestion
      evalTactic (← `(tactic| have%$tk $h : $a = ($val : $(← ppTerm ty)) := rfl))
    | some h, some (some _) =>
      evalTactic (← `(tactic| have%$tk $h : ($val : $(← ppTerm ty)) = $a := rfl))
```

File: Mathlib/Algebra/Category/GroupCat/Injective.lean [hunk 2]
```diff
@@ -156,4 +156,3 @@ instance injective_of_divisible [DivisibleBy A ℤ] :
 #align AddCommGroup.injective_of_divisible AddCommGroupCat.injective_of_divisible
 
 end AddCommGroupCat
```
Comment (alexjbest) [inline Mathlib/Algebra/Category/GroupCat/Injective.lean:158]:
```suggestion
end AddCommGroupCat

```

File: Mathlib/Tactic/Set.lean [hunk 3]
```diff
@@ -39,25 +39,26 @@ h2 : x = y
...
-  | some h, some none =>
-    evalTactic (← `(tactic| have $h : $a = $val := rfl))
-  | some h, some (some _) =>
-    evalTactic (← `(tactic| have $h : $val = $a := rfl))
-  | _, _ => pure ()
+    Term.addTermInfo' (isBinder := true) a (mkFVar fvar)
+    if rw.isNone then
+      evalTactic (← `(tactic| try rewrite [(id rfl : $val = $a)] at *))
+    let tt ← delab ty
+      (RBMap.empty.insert .root <| KVMap.empty.setBool `pp.structureInstanceTypes true)
```
Comment (alexjbest) [inline Mathlib/Tactic/Set.lean:58]:
```suggestion
      (RBMap.empty.insert .root <| KVMap.empty.setBool ((`pp).append `structureInstanceTypes) true)
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**